### PR TITLE
Add quantum phase estimation error

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
Adds method to compute the quantum phase estimation error from the spectral norm error of the input unitary operator.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by soranjh** [comment]
[sc-55443]

File: pennylane/templates/subroutines/qpe.py [hunk 1]
```diff
@@ -200,6 +200,30 @@ def estimation_wires(self):
...
+
+        >>> class CustomOP(qml.resource.ErrorOperation):
+        ...    @property
+        ...    def error(self):
+        ...       return 0.1
+        >>> Op = CustomOP(wires=[0])
+        >>> QPE = QuantumPhaseEstimation(Op, estimation_wires = range(1, 5))
+        >>> QPE.error
+        0.11010001000000001
+        """
```
Comment (trbromley) [inline pennylane/templates/subroutines/qpe.py:217]: So for my understanding, this is the operator error for QPE found by composing the errors of the `Op` unitary being applied? How is this related to the error one would get from QPE, i.e., the error in measuring the phase (which depends on things like the number of qubits in the phase estimation register)?

**Review by DSGuala — COMMENTED** [review]
Looks good to me. Just wondering whether we will have more than one error type inside QPE

File: pennylane/templates/subroutines/qpe.py [hunk 2]
```diff
@@ -200,6 +200,30 @@ def estimation_wires(self):
         """The estimation wires of the QPE"""
         return self._hyperparameters["estimation_wires"]
 
+    @property
+    def error(self):
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/templates/subroutines/qpe.py:204]: `spectral_norm_error` looks good to me.

File: pennylane/templates/subroutines/qpe.py (hunk 1 above)
Comment (soranjh) [reply pennylane/templates/subroutines/qpe.py:217]: > So for my understanding, this is the operator error for QPE found by composing the errors of the Op unitary being applied?

Yes. 

> How is this related to the error one would get from QPE, i.e., the error in measuring the phase (which depends on things like the number of qubits in the phase estimation register)?

This will be addressed in future extension that address the estimation error. This [reference](https://arxiv.org/abs/2111.10430) provides some insights on this.

File: pennylane/templates/subroutines/qpe.py (hunk 2 above)
Comment (Jaybsoni) [inline pennylane/templates/subroutines/qpe.py:204]: Also please inherit from the `ErrorOperation` class

**Review by Jaybsoni — COMMENTED** [review]
Left some comments. Missing changelog entry

File: tests/templates/test_subroutines/test_qpe.py [hunk 3]
```diff
@@ -38,6 +38,46 @@ def test_flatten_unflatten():
     assert op is not new_op
 
 
+def test_error_value():
+    """Test that QPE error is correct for a given unitary error."""
```
[the remark this answers is not in the document]
Comment (soranjh) [reply tests/templates/test_subroutines/test_qpe.py]: Updated.

File: pennylane/templates/subroutines/qpe.py [hunk 4]
```diff
@@ -18,9 +18,10 @@
 import pennylane as qml
 from pennylane.queuing import QueuingManager
 from pennylane.operation import AnyWires, Operation, Operator
+from pennylane.resource import ErrorOperation
 
 
-class QuantumPhaseEstimation(Operation):
+class QuantumPhaseEstimation(ErrorOperation, Operation):
```
Comment (DSGuala) [inline pennylane/templates/subroutines/qpe.py:24]: Does it need to inherit from both `ErrorOperation` _and_ `Operation`?

Comment (Jaybsoni) [reply pennylane/templates/subroutines/qpe.py:24]: Just one

Comment (Jaybsoni) [inline pennylane/templates/subroutines/qpe.py:24]:
```suggestion
class QuantumPhaseEstimation(ErrorOperation):
```

File: pennylane/templates/subroutines/qpe.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/templates/subroutines/qpe.py:204]: This is the error method inherited from the ErrorOperation class. It should not be renamed.

File: pennylane/templates/subroutines/qpe.py [hunk 5]
```diff
@@ -200,6 +201,31 @@ def estimation_wires(self):
...
+        >>> class CustomOP(qml.resource.ErrorOperation):
+        ...    @property
+        ...    def error(self):
+        ...       return 0.005
+        >>> Op = CustomOP(wires=[0])
+        >>> QPE = QuantumPhaseEstimation(Op, estimation_wires = range(1, 5))
+        >>> QPE.error()
+        0.075
+        """
+        unitary_error = self._hyperparameters["unitary"].error.error
```
Comment (Jaybsoni) [inline pennylane/templates/subroutines/qpe.py:218]: It is not always the case that the base "unitary" will have an error, in this case we assume it is exact and there is 0 error. Please add a check for this: 

```
base_unitary = self._hyperparameters["unitary"]
if not isinstance(base_unitary, ErrorOperation):
    return SpectralNormError(0.0) 
```

File: pennylane/templates/subroutines/qpe.py [hunk 6]
```diff
@@ -200,6 +201,31 @@ def estimation_wires(self):
...
+
+        sequence_error = qml.math.array(
+            [unitary_error * (2**i) for i in range(len(self.estimation_wires) - 1, -1, -1)],
+            like=qml.math.get_interface(unitary_error),
+        )
+
+        additive_error = qml.math.sum(sequence_error)
+
+        return qml.resource.error.SpectralNormError(additive_error)
```
Comment (Jaybsoni) [inline pennylane/templates/subroutines/qpe.py:227]: It should be safe to just import SpectralNormError at the top and using that

File: pennylane/templates/subroutines/qpe.py (hunk 5 above)
Comment (Jaybsoni) [reply pennylane/templates/subroutines/qpe.py:218]: Also please add the associated test which computes the error for an operator with no error

File: tests/templates/test_subroutines/test_qpe.py [hunk 7]
```diff
@@ -38,6 +38,74 @@ def test_flatten_unflatten():
...
+        m_apprx = qml.matrix(qml.QuantumPhaseEstimation(u_apprx, estimation_wires=range(1, 3)))
+
+        matrix_error = qml.math.max(qml.math.svd(m_exact - m_apprx, compute_uv=False))
+
+        operator = CustomOP(wires=[0])
+        qpe_error = qml.QuantumPhaseEstimation(operator, estimation_wires=range(1, 3)).error().error
+
+        assert np.allclose(qpe_error, matrix_error, atol=1e-4)
+
+    # @pytest.mark.all_interfaces
```
Comment (Jaybsoni) [inline tests/templates/test_subroutines/test_qpe.py:84]:
```suggestion
    @pytest.mark.all_interfaces
```

File: pennylane/templates/subroutines/qpe.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DSGuala) [reply pennylane/templates/subroutines/qpe.py:204]: Decided to keep it as `error` because the docstring is clear and it is a required `abstractmethod`.
Changes to support other types of errors can be supported by adding a keyword argument like `QPE.error(calculation='SpectralNorm')`

File: tests/templates/test_subroutines/test_qpe.py [hunk 8]
```diff
@@ -38,6 +38,74 @@ def test_flatten_unflatten():
...
+        assert np.allclose(qpe_error, expected_error)
+
+    def test_error_unitary(self):
+        """Test that QPE error is correct for a given unitary error."""
+
+        u_exact = qml.RY(0.50, wires=0)
+        u_apprx = qml.RY(0.51, wires=0)
+
+        class CustomOP(qml.resource.ErrorOperation):
+            @property
```
Comment (Jaybsoni) [inline tests/templates/test_subroutines/test_qpe.py:69]:
```suggestion
```

File: tests/templates/test_subroutines/test_qpe.py [hunk 9]
```diff
@@ -38,6 +38,74 @@ def test_flatten_unflatten():
...
+    @pytest.mark.parametrize(
+        # the reference error is computed manually for a QPE operation with 2 estimation wires
+        ("operator_error", "expected_error"),
+        [(0.01, 0.03), (0.02, 0.06), (0.03, 0.09)],
+    )
+    def test_error_operator(self, operator_error, expected_error):
+        """Test that QPE error is correct for a given custom operator."""
+
+        class CustomOP(qml.resource.ErrorOperation):
+            @property
```
Comment (Jaybsoni) [inline tests/templates/test_subroutines/test_qpe.py:53]:
```suggestion
```

File: tests/templates/test_subroutines/test_qpe.py [hunk 10]
```diff
@@ -38,6 +38,74 @@ def test_flatten_unflatten():
...
+        # the reference error is computed manually for a QPE operation with 2 estimation wires
+        ("operator_error", "expected_error"),
+        [(0.01, 0.03)],
+    )
+    @pytest.mark.parametrize("interface", ["autograd", "jax", "torch"])
+    def test_error_interfaces(self, operator_error, interface, expected_error):
+        """Test that the error method works with all interfaces"""
+
+        class CustomOP(qml.resource.ErrorOperation):
+            @property
```
Comment (Jaybsoni) [inline tests/templates/test_subroutines/test_qpe.py:95]:
```suggestion
```

File: pennylane/templates/subroutines/qpe.py (hunk 5 above)
Comment (Jaybsoni) [inline pennylane/templates/subroutines/qpe.py:218]:
```suggestion
        unitary_error = self._hyperparameters["unitary"].error.error()
```

**Review by DSGuala — APPROVED** [review]
Just left one comment about inheritance. Looks good from my side.

File: pennylane/templates/subroutines/qpe.py (hunk 5 above)
Comment (soranjh) [reply pennylane/templates/subroutines/qpe.py]: Added.

File: tests/templates/test_subroutines/test_qpe.py [hunk 11]
```diff
@@ -38,6 +38,81 @@ def test_flatten_unflatten():
...
+
+        matrix_error = qml.math.max(qml.math.svd(m_exact - m_apprx, compute_uv=False))
+
+        operator = CustomOP(wires=[0])
+        qpe_error = qml.QuantumPhaseEstimation(operator, estimation_wires=range(1, 3)).error().error
+
+        assert np.allclose(qpe_error, matrix_error, atol=1e-4)
+
+    @pytest.mark.all_interfaces
+    @pytest.mark.parametrize(
```
Comment (Jaybsoni) [inline tests/templates/test_subroutines/test_qpe.py:93]: Don't need to parametrize if there is only one option

File: pennylane/templates/subroutines/qpe.py [hunk 12]
```diff
@@ -200,6 +201,34 @@ def estimation_wires(self):
         """The estimation wires of the QPE"""
         return self._hyperparameters["estimation_wires"]
 
+    def error(self):
+        """The QPE error computed from the spectral norm error of the input unitary operator.
```
Comment (Jaybsoni) [inline pennylane/templates/subroutines/qpe.py:205]: We can come back to the documentation and add some more of the details when we advertise the feature

**Review by Jaybsoni — APPROVED** [review]
Looks good, a few small comments, nothing blocking
